Enemy Stat Randomizer added

Fantaselion opened the pull request:
I've added a fully functional Metroid Prime 1 Enemy Stat Randomizer, its in its own tab in the Game Modifications tab

I did have to change a couple things before publishing my PyPI package since the wheel for macOS wouldn't build unless I got rid of using sets 'since it doesn't think they exist apparently' so I hope I didn't break something (Probably didn't) but otherwise it works very well, should work across all game versions.

There is no disable button for my function but I made it so if all range values are set to default (1.00 - 1.00) then it just won't call my function, so basically the same thing.

toasterparty commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -60,6 +60,18 @@ class PrimeConfiguration(BaseConfiguration):
     qol_pickup_scans: bool
     qol_cutscenes: LayoutCutsceneMode
 
+    range_scale_low: float = dataclasses.field(metadata={"min": 0.01, "max": 25.0, "precision": 1.0})
```
please make these all part of 1 dict or class under 1 config e.g. `enemy_stat_rando: ...`

toasterparty reviewed (changes requested):
Please edit CHANGELOG.md to show your new feature. This file gets shown to the user when they update.

henriquegemignani commented on randovania/games/prime1/exporter/game_exporter.py:
```diff
@@ -197,3 +212,15 @@ def export_game(self, patch_data: dict, export_params: GameExportParams,
                 raise RuntimeError(f"randomprime panic: {e}") from e
+            
+
+        if all(v == 1.0 for v in [random_scale_min, random_scale_max,
```
Avoid having an empty if block with the logic on else. You can easily invert the criteria to `if any(v != 1.0 ...`

henriquegemignani commented on randovania/games/prime1/exporter/game_exporter.py:
```diff
@@ -7,6 +7,8 @@
 
 import py_randomprime
 
+import PARAMETEREDITOR.PARAMETEREDITOR
```
That's a really weird name module name you used for the package.

Fantaselion replied:
ngl if I could go back and tell myself to change it to something better I would....idk what I was thinking.

toasterparty commented:
It looks good to me. @henriquegemignani gets final say

Fantaselion commented on setup.cfg:
```diff
@@ -46,6 +46,7 @@ install_requires =
     open-prime-rando>=0.1.0
     cave-story-randomizer>=1.0.0
     tsc-utils>=0.2.0
+	parametereditor>=0.2.1
```
Yea that was an mistake

henriquegemignani reviewed (changes requested):
UI feedback.

* Call it `Enemy Attribute` instead of Stat, and remove the Rando.
* Move this tab to before `Other`, that one makes sense being always last.
* Remove the group box, it's redundant with the title of the tab.
* Use some kind of layout for positioning the elements, like QGridLayout, to position the elements dynamically. 
* Be considerably more descriptive on the text. The description should also be readable by someone with no technical knowledge at all or someone not familiar with randomizers, so you should stick to very simple terms as much as possible.
* This should mention that these changes are not logical at all.

Fantaselion commented:
> UI feedback.
> 
> * Call it `Enemy Attribute` instead of Stat, and remove the Rando.
> * Move this tab to before `Other`, that one makes sense being always last.
> * Remove the group box, it's redundant with the title of the tab.
> * Use some kind of layout for positioning the elements, like QGridLayout, to position the elements dynamically.
> * Be considerably more descriptive on the text. The description should also be readable by someone with no technical knowledge at all or someone not familiar with randomizers, so you should stick to very simple terms as much as possible.
> * This should mention that these changes are not logical at all.

Done! hope my description isn't too long now.

henriquegemignani commented on randovania/games/prime1/exporter/game_exporter.py:
```diff
@@ -197,3 +212,15 @@ def export_game(self, patch_data: dict, export_params: GameExportParams,
+           PARAMETEREDITOR.PARAMETEREDITOR.PyPARAMETEREDITOR(new_config["inputIso"], new_config["outputIso"], patch_data["seed"], random_scale_min, random_scale_max, random_health_min,
+                                              random_health_max, random_speed_min, random_speed_max, random_damage_min,
+                                              random_damage_max, random_knockback_min, random_knockback_max, diff_XYZ) 
+        else:
```
this else is not needed

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,31 @@ def describe_probability(probability, attribute):
+                rand_range[0] = rand_range[1]
+                rand_range[1] = temp
+
+            return "randomized " + attribute + " within range " + str(rand_range[0]) + " - " + str(rand_range[1])
```
Just `Random ` instead of `Randomized`.
You can also use an fstring here to build this text.

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,31 @@ def describe_probability(probability, attribute):
+                return None
+            elif rand_range[0] > rand_range[1]:
+                temp = rand_range[0]
+                rand_range[0] = rand_range[1]
```
`rand_range = (rand_range[1], rand_range[0])`

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,31 @@ def describe_probability(probability, attribute):
+            elif diff_xyz == True:
+                return "Scale XYZ values will be randomized separately"
+            elif diff_xyz == False:
+                return "Scale XYZ values will be randomized together"
```
This is the default behavior, so we don't need to mention this case

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,31 @@ def describe_probability(probability, attribute):
+            if enemy_rando_range_scale is None:
+                return None
+            elif diff_xyz == True:
+                return "Scale XYZ values will be randomized separately"
```
Thinking about this, calling this case as just "stretching" (or something similar) works great.
Talking about changing x/y/z separately is not intuitive if you're not familiar to how models can be scaled.

henriquegemignani reviewed (changes requested):
The new description is a lot better! But I feel it doesn't flow great, but my skills in revising text are bad. Take a picture of the screen and ask around for help.

The layout is almost there. There's some significant borders to the left and right which, while fine, are very inconsistent with all the other tabs. In particular, take a look at the Elevators tab.

Fantaselion commented:
Updated my fork to work with the new Randovania version, also I was able to get 'sets' working on the mac-OS Wheel so the package this time is the 'tested' version (as before I didn't re-test when I switched from sets to something else)

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,28 @@ def describe_probability(probability, attribute):
+        def different_xyz_randomization(diff_xyz):
+            if enemy_rando_range_scale is None:
+                return None
+            elif diff_xyz == True:
```
Just `elif diff_xyz:`

henriquegemignani commented on requirements.txt:
```diff
@@ -160,6 +160,8 @@ packaging==21.3
     # via
     #   matplotlib
     #   pytest
+parametereditor==2.1.1
```
Inconsistent with the `parametereditor>=2.1.2` in setup.cfg

henriquegemignani commented on test/games/prime1/gui/preset_settings/test_prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,26 @@
+from randovania.interface_common.preset_editor import PresetEditor
+from randovania.games.game import RandovaniaGame
+
+@pytest.mark.parametrize("game", [RandovaniaGame.METROID_PRIME])
```
unused parameter

henriquegemignani reviewed (changes requested):
Can you merge main and resolve conflicts, please?

henriquegemignani commented on randovania/games/prime1/exporter/game_exporter.py:
```diff
@@ -197,3 +212,13 @@ def export_game(self, patch_data: dict, export_params: GameExportParams,
+                                  random_speed_min, random_speed_max,
+                                  random_damage_min, random_damage_max,
+                                  random_knockback_min, random_knockback_max]):
+            PARAMETEREDITOR.PARAMETEREDITOR.PyPARAMETEREDITOR(new_config["inputIso"], new_config["outputIso"], patch_data["seed"], random_scale_min, random_scale_max, random_health_min,
```
Are you overwriting outputIso with a copy of inputIso? This will throw away any changes made by randomprime.

toasterparty replied:
randomprime changes are not being clobbered

henriquegemignani commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -60,6 +60,18 @@ class PrimeConfiguration(BaseConfiguration):
     qol_pickup_scans: bool
     qol_cutscenes: LayoutCutsceneMode
 
+    enemy_rando_range_scale_low: float = dataclasses.field(metadata={"min": 0.01, "max": 25.0, "precision": 1.0})
```
There should be a dangerous setting entry if any of these is modified.

toasterparty reviewed (changes requested):
- Your program hangs forever if the input ISO is prime practice mod (legacy)
- Your program hangs forever is the output game format is `.ciso` or `.gcz`
- When a player in a multiworld session has enemy stat rando enabled, the person generating the seed must get a warning that the seed might be incompletable due to dangerous settings
- Does it work on NTSC 0-02 PAL etc? Last I heard it was fixed offset patching. If it doesn't support certain version, you need to have a friendly message interrupt the user. UltiNaruto has every version of the game if you need someone to test something
- If my boss size randomizer is enabled in addition to your enemy size randomizer, disable my randomizer so that they do not stack

Fantaselion commented:
> * Your program hangs forever if the input ISO is prime practice mod (legacy)
> * Your program hangs forever is the output game format is `.ciso` or `.gcz`
> * When a player in a multiworld session has enemy stat rando enabled, the person generating the seed must get a warning that the seed might be incompletable due to dangerous settings
> * Does it work on NTSC 0-02 PAL etc? Last I heard it was fixed offset patching. If it doesn't support certain version, you need to have a friendly message interrupt the user. UltiNaruto has every version of the game if you need someone to test something
> * If my boss size randomizer is enabled in addition to your enemy size randomizer, disable my randomizer so that they do not stack

> * I'll try, will most likely be an easy fix
> * How does the ciso format know where the pak locations are? I need that info to be able to work with it, also the gcz is compressed and I can't work with it at all since I have no code to parse anything so the only way my code would work with gcz is if I edited a copy of the input file before randomprime gets it, that way i'm not "reading" from the output file (turns out I don't even need the input file right now)
> * How do I set that? I don't remember seeing that anywhere
> * I've changed my code to combine my offset searcher and my stat randomizer due to what darkzero said about it being a bad idea to just use plain offsets so assuming the offsets for the pointers to the pak locations are in the same place, it should work, (if not I can just have different offsets per version) will ask UltiNaruto though
> * Will do

toasterparty commented:
@Fantaselion to clarify, you don't NEED to support ciso/gcz/practice-mod/PAL etc.. your program just needs to not explode when these things happen. 

this:
> When a player in a multiworld session has enemy stat rando enabled, the person generating the seed must get a warning that the seed might be incompletable due to dangerous settings

and Darkszero's comment are referring to the same thing

Fantaselion commented:
- After testing and code changes/fixes, my program will no longer hang if something goes wrong, it will automatically abort with an exception and not crash randovania.

- My code now works with all 6 Metroid Prime GC versions, but outputting as a ciso, or gcz will cause an exception and abort my code.

- I put a warning about possible incompatibility with multiworld in the UI, while also mentioning that boss size randomizer will be turned off(I also made sure that the preset describer won't say random Boss Sizes if my scale is changed) if any of my min-max scales are changed. (also mentioned that they have to set output game type to iso)

- I added my values to dangerous settings and then added those to the RDVgame (prime1_crazy_seed) and the corresponding test file but its always failing. My values aren't updated from default to what I put in the RDVgame so idk what i'm doing wrong.

henriquegemignani commented on tools/.gitignore:
```diff
@@ -1,12 +1,85 @@
-/Randomizer.zip
```
Please undo the changes to this file.
It seems you copied over the gitignore in the repository root to here.

Fantaselion replied:
Hmm it seems pulling created some issues, hopefully nothing else got messed up.

Fantaselion commented:
So I started working on putting my values into a class in prime configuration, and trying to get this to work, but i'm just getting confused lol, so need some input on this. (just commited what i've done so far)

henriquegemignani commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -21,6 +21,20 @@ class LayoutCutsceneMode(BitPackEnum, Enum):
     MAJOR = "major"
 
+class EnemyAttributeRandomizer(BitPackDataclass):
+    optional_int: int | None = dataclasses.field(metadata={"min": 0, "max": 15})
```
I don't see a purpose for this?

Fantaselion replied:
Whoops, I copied the values from test for reference but I forgot to delete it.

henriquegemignani commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -21,6 +21,20 @@ class LayoutCutsceneMode(BitPackEnum, Enum):
     MINOR = "minor"
     MAJOR = "major"
 
+class EnemyAttributeRandomizer(BitPackDataclass):
```
Missing `@dataclasses.dataclass(frozen=True)` above the class definition

Fantaselion commented:
So more confusion, if I set the "enemy_attributes" to None in 'preset_migration', then it will delete all the fields so I have to re-add them in 'prime_enemy_stat_randomizer.py' but I can't since its a NoneType...So do I change the type back to the class type I put in prime_configuration? But then what was the point of having it be None to begin with? I could be thinking about this completely wrong though...

Fantaselion commented:
Ok I got everything working, all my variables are in a class and are nice, clean, and compact. I also updated my tab to work with the recent randovania changes to them.

henriquegemignani commented on randovania/games/prime1/exporter/game_exporter.py:
```diff
@@ -197,3 +202,19 @@ def export_game(self, patch_data: dict, export_params: GameExportParams,
                 raise RuntimeError(f"randomprime panic: {e}") from e
+
+        if random_enemy_attributes:
+            rand_scale_low = random_enemy_attributes.enemy_rando_range_scale_low
```
Everything in newConfig is dicts, so this should be accessing via `[]`.
If no test failed because of it, then we need improvements to the tests

Fantaselion replied:
It gives me "EnemyAttributeRandomizer" object is not subscriptable.

henriquegemignani commented on randovania/games/prime1/exporter/patch_data_factory.py:
```diff
@@ -796,7 +796,7 @@ def component_number(name):
 
         boss_sizes = None
-        if self.configuration.random_boss_sizes:
+        if self.configuration.random_boss_sizes and all(v == 1.0 for v in [self.configuration.enemy_rando_range_scale_low, self.configuration.enemy_rando_range_scale_high]):
```
Checking fields of config that no longer exist. It should also just check if the enemy-attributes exists

henriquegemignani commented on randovania/games/prime1/exporter/patch_data_factory.py:
```diff
@@ -822,6 +822,7 @@ def get_random_size(minimum, maximum):
                 "platedBeetle": get_random_size(0.05, 6.0),
                 "cloakedDrone": get_random_size(0.05, 6.0),  # only scales width (lmao)
             }
+            print("YO")
```
Prints

Fantaselion replied:
XD whoops, I thought I cleaned all these up.

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,188 @@
+                                                                                            self.diff_xyz.isChecked()))
+            else:
+                editor.set_configuration_field("enemy_attributes", None)
+        self.minimum_label.setEnabled(checked)
```
You could use an array of all widgets and a for loop

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,188 @@
+    def on_preset_changed(self, preset: Preset):
+        config = preset.configuration
+        if config.enemy_attributes:
+            with self._editor as editor:
```
You don't need the editor

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,188 @@
+
+    def on_preset_changed(self, preset: Preset):
+        config = preset.configuration
+        if config.enemy_attributes:
```
Use explicit `is not None` checks

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,188 @@
+        self.label_2.setEnabled(checked)
+
+    def on_preset_changed(self, preset: Preset):
+        config = preset.configuration
```
You should set the state of `self.Activate_Randomizer` accordingly here too

henriquegemignani replied:
You're missing setting to `False` when enemy_attributes  is None

Fantaselion replied:
Thats weird, I did that in my copy but it didn't commit that....but it commited mostly everything else?

henriquegemignani replied:
You're still missing a `self.activate_randomizer.setChecked(False)` in the else case

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,188 @@
+        with self._editor as editor:
+            config = editor.configuration.enemy_attributes
+            config = dataclasses.replace(config, enemy_rando_range_scale_high=self.range_scale_high.value())
+            editor.set_configuration_field(
```
You can minimise repeated code by making a helper function that takes a Enemy Attributes object and does the with editor set configuration field stuff

henriquegemignani commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -88,6 +105,13 @@ def dangerous_settings(self) -> list[str]:
         if self.allow_underwater_movement_without_gravity:
             result.append("Dangerous Gravity Suit Logic")
 
+        if any(v != 1.0 for v in [self.enemy_rando_range_scale_low, self.enemy_rando_range_scale_high,
```
Check if attributes is not None instead of this

Fantaselion commented:
Need help with fixing the test for dangerous settings.

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/__init__.py:
```diff
@@ -5,6 +5,9 @@
 
 
 def prime1_preset_tabs(editor: PresetEditor, window_manager: WindowManager):
+    game_enum = editor.game
```
these lines are not needed anymore

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,147 @@
+            self.diff_xyz.setChecked(config.enemy_attributes.enemy_rando_diff_xyz)
+
+    def _on_spin_changed_range_scale_low(self):
+        with self._editor as editor:
```
the with block is not needed since it's in `_set_up_config`

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,147 @@
+            config = dataclasses.replace(config, enemy_rando_diff_xyz=self.diff_xyz.isChecked())
+            self._set_up_config(config)
+
+    def _set_up_config(self, config):
```
This name is not that great, maybe something like `_set_enemy_attributes_in_editor`?

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -44,6 +44,35 @@ def describe_probability(probability, attribute):
+            elif diff_xyz:
+                return "Enemies will be stretched randomly"
+
+        if configuration.enemy_attributes:
```
use `is not None`

henriquegemignani commented on randovania/games/prime1/layout/preset_describer.py:
```diff
@@ -111,9 +140,21 @@ def describe_probability(probability, attribute):
                 }
             ],
         }
+        if enemy_rando_range_scale != None:
```
`is not None`

Fantaselion commented:
test_prime_game_exporter and test_prime_dangerous_settings are giving me new errors I don't know how to solve.

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,136 @@
+            self.diff_xyz.setChecked(config.enemy_attributes.enemy_rando_diff_xyz)
+
+    def _on_spin_changed_range_scale_low(self):
+            config = self._editor.configuration.enemy_attributes
```
there's one extra level of indentation on all these functions

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,136 @@
+        self.enemy_stat_randomizer_description.setText(self.enemy_stat_randomizer_description.text().replace("color:#0000ff;", ""))
+
+        # Signals
+        self.Activate_Randomizer.stateChanged.connect(self._on_activation)
```
Please rename this widget to follow snake case

henriquegemignani commented on randovania/games/prime1/layout/prime_configuration.py:
```diff
@@ -88,6 +105,9 @@ def dangerous_settings(self) -> list[str]:
         if self.allow_underwater_movement_without_gravity:
             result.append("Dangerous Gravity Suit Logic")
 
+        if enemy_attributes is not None:
```
Missing `self`

henriquegemignani commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -15,7 +16,9 @@ def test_patch_game(mocker, tmp_path, use_echoes_models):
-    patch_data = {"patch": "data", 'gameConfig': {}, 'hasSpoiler': True, "preferences": {}, "roomRandoMode": "None"}
+    patch_data = {"patch": "data", 'gameConfig': {}, 'hasSpoiler': True, "preferences": {}, "roomRandoMode": "None",
+                  "randEnemyAttributes": EnemyAttributeRandomizer(1.1, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, False).as_json,
+                  "seed": 106827108}
```
There's a mismatch in parameters actually sent to `randomprime` and the ones we expected to be. The actual ones has the `seed`.

Fantaselion replied:
If I get rid of the "seed" parameter then I get "seed" is not defined when my function is called, so how do I access the actual seed?

Fantaselion replied:
Also my enemy_attributes keeps being sent to "dangerous_settings" as None rather than using what I have here in patch_data

henriquegemignani commented on .idea/.name:
```diff
@@ -1 +1 @@
-randovania
```
Please don't change these `.idea` files.

FYI, if you're using PyCharm it _should_ be possible to open the root folder as a project

Fantaselion replied:
How in the world di that happen? I didn't even open that and it didn't show up when I typed "git status"

henriquegemignani commented on randovania/games/prime1/gui/preset_settings/prime_enemy_stat_randomizer.py:
```diff
@@ -0,0 +1,136 @@
+        config = dataclasses.replace(config, enemy_rando_range_knockback_high=self.range_knockback_high.value())
+        self._set_enemy_attributes_in_editor(config)
+
+    def _on_check_change(self):
```
This name is fairly generic, add a reference to diff_xyz

henriquegemignani commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -15,7 +16,8 @@ def test_patch_game(mocker, tmp_path, use_echoes_models):
     mock_patch_iso_raw: MagicMock = mocker.patch("py_randomprime.patch_iso_raw")
```
You should also mock your patcher module and assert that it is being called with the values you expect. You can use this line here as a reference.

Ideally, you'd also add another parametrize to have a test with the enemy attributes being enabled or disabled. Should be similar to how `use_echoes_models` is done.

henriquegemignani commented on test/test_files/log_files/prime1_crazy_seed.rdvgame:
```diff
@@ -1251,7 +1251,8 @@
                     "qol_pickup_scans": true,
-                    "qol_cutscenes": "major"
+                    "qol_cutscenes": "major",
+					"enemy_attributes": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, false]
```
Did you also bump the preset version?

henriquegemignani commented on randovania/games/prime1/exporter/patch_data_factory.py:
```diff
@@ -902,6 +904,8 @@ def get_random_size(minimum, maximum):
             "hasSpoiler": self.description.has_spoiler,
             "roomRandoMode": self.configuration.room_rando.value,
 
+            "randEnemyAttributes": self.configuration.enemy_attributes and self.configuration.enemy_attributes.as_json
```
`self.configuration.enemy_attributes.as_json if self.configuration.enemy_attributes is not None else None,`

henriquegemignani commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -5,17 +5,26 @@
         "UpdateHintState__13CStateManagerFf": 0x80044D38,
     })
+    if use_enemy_attribute_randomizer:
+        use_enemy_attribute_randomizer = EnemyAttributeRandomizer(1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, False).as_json
```
use a different variable here, it's less confusing

henriquegemignani commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -67,6 +76,11 @@ def test_patch_game(mocker, tmp_path, use_echoes_models):
         assert not asset_cache_path.exists()
 
+    if use_enemy_attribute_randomizer:
+        mock_enemy_data.assert_called_once_with("inputIso", "outputIso", "seed", 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, False)
```
`tmp_path.joinpath("input.iso")` instead of `"inputIso"`
`tmp_path.joinpath("output.iso")` instead of `"outputIso"`
whatever you pass as a seed instead of `"seed"`

henriquegemignani commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -5,17 +5,26 @@
     mocker.patch("randovania.games.prime1.exporter.game_exporter.adjust_model_names")
-    patch_data = {"patch": "data", 'gameConfig': {}, 'hasSpoiler': True, "preferences": {}, "roomRandoMode": "None"}
+    patch_data = {"patch": "data", 'gameConfig': {}, 'hasSpoiler': True, "preferences": {}, "roomRandoMode": "None",
+                  "randEnemyAttributes": use_enemy_attribute_randomizer, "seed": 103817502} # I still have the "seed" here because if I don't have this here, my function creates a  KeyError: 'seed'
```
you should store the seed number in a variable.

What you're doing wrong with the seed is that your code isn't poping it from the config, like you do with `"randEnemyAttributes"`. So it stays into what is sent to randomprime. You either should pop it so it's not sent, or update  the `expected` dict below to properly expect this value.

Fantaselion commented:
So at this point I get 
```
FAILED test_prime_game_exporter.py::test_patch_game[True-True] - AssertionError: Expected 'PyRandom_Enemy_Attributes' to be called once. Called 0 times.
FAILED test_prime_game_exporter.py::test_patch_game[True-False] - AssertionError: Expected 'PyRandom_Enemy_Attributes' to be called once. Called 0 times.
```
I know my code is being called as I get the expected cout from my wrapped function 
```
Couldn't find output file.
Aborting Enemy Stat Randomizer
```
and the call to my function literally is litteraly `PyRandom_Enemy_Attributes(stuff)` so idk why its saying its not being called.

henriquegemignani commented:
When you do `mocker.patch`, you're replacing the original function with a "mock" that does nothing but keep track of how it was called. The fact that the prints from your patcher are being called implies it's not been properly replaced with a mock.

The reason is a bit tricky and related to how the import for it is done. You've used `from Random_Enemy_Attributes.Random_Enemy_Attributes import PyRandom_Enemy_Attributes`. Try going back to just `import Random_Enemy_Attributes.Random_Enemy_Attributes`.

Alternative, you can try `mocker.patch("randovania.games.prime1.exporter.game_exporter.PyRandom_Enemy_Attributes")`

Fantaselion commented:
Phew I finally fixed all exporter tests. Idk if there will be any issues with other tests but its 7:12 A.M. and I need sleep lol.

henriquegemignani commented on test/test_files/log_files/prime1-vanilla.rdvgame:
```diff
@@ -176,7 +176,17 @@
                     "qol_minor_cutscenes": false,
-                    "qol_major_cutscenes": false
+                    "qol_major_cutscenes": false,
+					"enemy_rando_range_scale_low": 1.00,
```
These changes shouldn't be here

henriquegemignani reviewed (approved):
Pending a review from @toasterparty

Fantaselion commented:
All my new code should now be covered by tests!

toasterparty commented on test/test_files/randomprime_expected_data_one_way_door.json:
```diff
@@ -34,24 +34,8 @@
-            "cloakedDrone": 0.6282245815635213
-        },
+        "resultsString": "4.6.0.dev146 | Seed Hash - Geemer Shriekbat Fission (AAAAAAAA)",
+        "bossSizes": null,
```
The preset corresponding to this `json` output specifies:
```json
"random_boss_sizes": true,
```

It does NOT specify random attributes.

The expected behavior is that a random boss size dict is placed here.

toasterparty commented on test/test_files/randomprime_expected_data_one_way_door.json:
```diff
@@ -34,24 +34,8 @@
-            "cloakedDrone": 0.6282245815635213
-        },
+        "resultsString": "4.6.0.dev146 | Seed Hash - Geemer Shriekbat Fission (AAAAAAAA)",
+        "bossSizes": null,
         "noDoors": true,
```
The preset corresponding to this `json` output specifies:
```json
"random_boss_sizes": true,
```

It does NOT specify random attributes.

The expected behavior is that a random boss size dict is placed here.

Fantaselion replied:
I asked DarkZero how to handle changing this based on if I'm randomizing scale, and he told me that its difficult to do that, but what I could do is change it in patch_data_factory, so I changed your
`if self.configuration.random_boss_sizes` to
`if self.configuration.random_boss_sizes and self.configuration.enemy_attributes is not None:`
I set up preset_describer to take this in account also.
Though your right, this makes the .rdvgame and .json file not say the right thing....Idk a better way to do this.

toasterparty commented on test/games/prime1/exporter/test_prime_game_exporter.py:
```diff
@@ -5,19 +5,27 @@
     })
+    seed = 103817502
+    if use_enemy_attribute_randomizer:
+        enemy_attribute_randomizer = EnemyAttributeRandomizer(1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, False).as_json
```
Why are you testing with values that do nothing? Shouldn't you test with values that actually make game modifications?

Fantaselion replied:
Ahh right, was focused on the other aspects of the test.
